src/bech32.cpp

```diff
@@ -161,18 +161,25 @@ std::pair<std::string, data> Decode(const std::string& str) {
     for (size_t i = 0; i < str.size(); ++i) {
         unsigned char c = str[i];
         if (c < 33 || c > 126) return {};
-        if (c >= 'a' && c <= 'z') lower = true;
-        if (c >= 'A' && c <= 'Z') upper = true;
```

Comment:
`else` here would improve a bit.

Reply:
Maybe better:
```cpp
if (c >= 'a') { if (c <= 'z') lower = true; }
else if (c >= 'A') { if (c <= 'Z') upper = true; }

Reply:
@promag Do you have an example in mind of a most common case?

Reply:
The success case?

Reply:
OK, in that case (let's assume a string consisting only of lowercase or uppercase letters) the existing code will do three comparisons and one assignment, per character. The proposed code will do three comparisons per character, with the added benefit of returning earlier in the case of a malformed string. Not sure why you would "suspect this makes the most common case worst".

Reply:
Further to this, if we're going to assume the most common case is the success case (which is reasonable), it would probably be good to move the initial (c < 33 || c > 126) check to an ultimate "else if" check to the original code.

Reply:
@murrayn probably, but can you post benchmark results?